### PR TITLE
fix(gateway): stop reconnecting on client errors

### LINKED ISSUE #2448: Gateway should not reattempt portal connection after 4xx
```
2023-10-19T20:13:08.626448Z  WARN connlib_shared: get_device_id() couldn't load SMBios. Error: No such file or directory (os error 2)
2023-10-19T20:13:08.626552Z  WARN connlib_shared: get_device_id() couldn't find a SMBios Serial. Using random UUIDv4 instead.
2023-10-19T20:13:08.719711Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=731.958823ms
2023-10-19T20:13:09.480553Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=491.139639ms
2023-10-19T20:13:09.993478Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=1.54540021s
2023-10-19T20:13:11.566011Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=2.242893257s
2023-10-19T20:13:13.833139Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=2.846466541s
2023-10-19T20:13:16.704676Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=3.942577036s
2023-10-19T20:13:20.669871Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=4.47657713s
2023-10-19T20:13:25.166530Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=8.235372903s
2023-10-19T20:13:33.429449Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=7.298842497s
2023-10-19T20:13:40.754268Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=20.145194971s
```

### PR DESCRIPTION
**Pull request by thomaseizinger**
Fixes: #2448.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
Note: I haven't tested this manually because I wasn't sure how to do that.

**Review by jamilbk — COMMENTED** [review]
This LGTM. We should probably apply the same treatment to the Relay?

I'll defer to @conectado for approval

**Comment by conectado** [comment]
I tested the PR by:
* Changing the `PORTAL_TOKEN` to garbage in the gateway section `docker-compose.yml`
* `docker compose up`

And it doesn't seem to be working:

```
firezone-review-gateway-1  | 2023-10-20T20:21:39.433538Z  WARN connlib_shared: get_device_id() couldn't load SMBios. Error: No such file or directory (os error 2)
firezone-review-gateway-1  | 2023-10-20T20:21:39.433550Z  WARN connlib_shared: get_device_id() couldn't find a SMBios Serial. Using random UUIDv4 instead.
firezone-review-gateway-1  | 2023-10-20T20:21:39.444828Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=738.023676ms
firezone-review-gateway-1  | 2023-10-20T20:21:40.188341Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=464.430746ms
firezone-review-gateway-1  | 2023-10-20T20:21:40.657249Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=586.104133ms
firezone-review-gateway-1  | 2023-10-20T20:21:41.247497Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=1.043084023s
firezone-review-gateway-1  | 2023-10-20T20:21:42.298364Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=1.291970354s
firezone-review-gateway-1  | 2023-10-20T20:21:43.598687Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=5.148894118s
firezone-review-gateway-1  | 2023-10-20T20:21:48.758225Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=3.311668991s
firezone-review-gateway-1  | 2023-10-20T20:21:52.079878Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=9.187699153s
firezone-review-gateway-1  | 2023-10-20T20:22:01.278015Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=15.499256704s
firezone-review-gateway-1  | 2023-10-20T20:22:16.787930Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=12.509658101s
firezone-review-gateway-1  | 2023-10-20T20:22:29.306218Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=20.509953989s
firezone-review-gateway-1  | 2023-10-20T20:22:49.823802Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=57.155158121s
firezone-review-gateway-1  | 2023-10-20T20:23:46.989198Z  WARN firezone_gateway: Error connecting to portal: HTTP error: 401 Unauthorized retry_in=64.67684299s
```

File: rust/gateway/src/main.rs [hunk 1]
```diff
@@ -80,3 +98,12 @@ struct Cli {
     #[command(flatten)]
     common: CommonArgs,
 }
+
+/// Checks whether the given [`std::error::Error`] is in-fact an HTTP error with a 4xx status code.
+fn is_client_error(e: &(dyn std::error::Error + 'static)) -> bool {
+    let Some(tungstenite::Error::Http(r)) = e.downcast_ref() else {
```
Comment (conectado) [inline rust/gateway/src/main.rs:104]:
```suggestion
    let Some(phoenix_channel::Error::WebSocket(tungstenite::Error::Http(r))) = e.downcast_ref() else {
```

Seems like we wrapped `tungstenite`'s error with one from `phoenix_channel`, so this downcast failed and we were returning a `Transient` error. 

I tested with this change and it works :D

**Review by conectado — CHANGES_REQUESTED** [review]

**Comment by jamilbk** [comment]
Merged in main to update required checks

File: rust/gateway/src/main.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/gateway/src/main.rs:104]: Hmm, it looks like we are not exposing the error correctly as `source()` then.

Thanks for the pointers in testing this, will iterate on it!

**Comment by thomaseizinger** [comment]
> I tested the PR by:
> 
>     * Changing the `PORTAL_TOKEN` to garbage in the gateway section `docker-compose.yml`
> 
>     * `docker compose up`

Retested this and is working now!

**Review by conectado — APPROVED** [review]

**Pull request merged**